enigma-principal/enclave/src/epoch_keeper_t/epoch_t.rs

```diff
@@ -28,12 +30,27 @@ impl Epoch {
 impl RawEncodable for Epoch {
     /// Encode the Epoch as Ethereum ABI parameters
     fn raw_encode(&self) -> Bytes {
-        let tokens = vec![
-            Token::Uint(self.seed),
-            Token::Uint(self.nonce),
-            Token::Array(self.worker_params.workers.iter().map(|a| Token::Address(*a)).collect()),
-            Token::Array(self.worker_params.stakes.iter().map(|s| Token::Uint(*s)).collect()),
-        ];
-        encode(&tokens)
+        let raw_seed = H256::from(self.seed).0.to_vec();
```

Comment:
Can you please explain what are you doing here exactly? why don't use use the regular `hash_multiple`? Implementing the same thing multiple times is a *bad* idea.
(especially something as sensitive as this)

Reply:
@elichai @fredfortier 
Definitely KM node should use a general core implementation of hashing for multiple values, and not use a special function for specific values. The problem is that today in core there is no correct implementation of hashing of multiple values that may include nested arrays. Also the @fredfortier's implementation of `raw_encode` is not good as it may lead to the same output for different inputs. I and @elichai found an encoding for nested arrays, for which I did a formal proof that for any two different inputs, the outputs will be always different: https://github.com/enigmampc/protocol-discovery/blob/master/docs/hash_mul_nested.pdf. 
So it seems to me that:
 1. The correct implementation should be added to core
 2. KM node should use this implementation
 3. The implementation of hashing in the contract for signature verification should be changed accordingly.
Meanwhile before 1 is done, I am OK to leave `raw_encode` as is, because signature verification works now, and we can progress with integration tests and other stuff. Once 1 is done, we will do 2 and 3 ASAP.

Reply:
> Also the @fredfortier's implementation of raw_encode is not good as it may lead to the same output for different inputs.

@moriaab Do you mind providing an example? This isn't my implementation, I implemented exactly what I was told the algorithm was in order to match the contract.